nexus/reconfigurator/planning/src/example.rs

```diff
@@ -4,139 +4,267 @@
+    where
+        F: FnOnce(SledBuilder) -> SledBuilder,
+    {
+        self.invalidate_caches();
+
+        let sled_id = self.sled_rng.next();
+        let sled = SledBuilder::new().id(sled_id);
+        let sled = f(sled);
+        let _ = self.system.sled(sled).expect("new sled added successfully");
+        sled_id
+    }
+
+    /// Sets the Omicron zones for a sled, invalidating internal caches.
+    pub fn sled_set_omicron_zones(
+        &mut self,
+        sled_id: SledUuid,
+        zones: OmicronZonesConfig,
+    ) -> anyhow::Result<()> {
+        self.invalidate_caches();
+        self.system.sled_set_omicron_zones(sled_id, zones)?;
+
+        Ok(())
+    }
+
+    /// Returns the planning input for the system.
+    ///
+    /// This is consistent with the system description, and is regenerated if
+    /// [`Self::system_mut`] is called.
+    pub fn planning_input(&self) -> &PlanningInput {
+        &self.get_or_init_derived().input
+    }
+
+    /// Returns an inventory collection representing the system.
+    ///
+    /// This is consistent with the system description, and is regenerated if
+    /// [`Self::system_mut`] is called. Some tests may wish to simulate
+    /// situations where inventory collection lags a bit -- in those cases, it
+    /// is recommended that you clone the collection.
+    pub fn collection(&self) -> &Collection {
+        &self.get_or_init_derived().collection
+    }
+
+    fn get_or_init_derived(&self) -> &ExampleSystemDerived {
+        self.derived.get_or_init(|| {
+            let input_builder = self
+                .system
+                .to_planning_input_builder()
+                .expect("failed to make planning input builder");
+            let input = input_builder.build();
+
+            let collection_builder = self
+                .system
+                .to_collection_builder()
+                .expect("failed to build collection");
+            let mut collection = collection_builder.build();
+            // This is a somewhat gross hack.
+            collection.id = self.next_collection_id;
+
+            ExampleSystemDerived { input, collection }
+        })
+    }
+
+    // This must be called whenever the system description changes.
+    fn invalidate_caches(&mut self) {
+        // If the derived cache wasn't created, then we don't need to
+        // invalidate anything (and in particular we don't need to bump the
+        // collection ID). Note that this is a `&mut self` method, which means
+        // that we're statically guaranteed that nothing else is attempting to
+        // initialize derived.
+        if self.derived.get().is_none() {
+            return;
         }
+        self.derived = OnceLock::new();
+        self.next_collection_id = self.collection_rng.next();
+        slog::debug!(
+            self.log,
+            "invalidated derived cache";
+            "next_collection_id" => %self.next_collection_id,
+        );
```

Comment:
This code implements the planning input and collection as derived from the SystemDescription, invalidated if the description changes.

Reply:
(this is moot for now, will be moved into its own PR)